![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Sebastián Tinoco, Stefano Schiappacasse
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Valentina Zuñiga, Ángelo Muñoz 

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: J. Vicente Onetto
- Estudiante n°2: Renzo Zanca

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

---

## 📖 Enunciado 

<div align="center">
    <img src='https://github.com/MDS7202/MDS7202/blob/main/recursos/2025-01/proyecto/proyecto.png?raw=true' style="border-radius: 12px"> 
</div>

En el competitivo universo de las bebidas gaseosas, la empresa **SodAI Drinks 🥤** ha logrado destacarse por su creatividad, diversidad de productos y enfoque centrado en el cliente. Ofrece una extensa gama de bebidas carbonatadas que abarca distintos segmentos del mercado: desde productos premium en presentaciones sofisticadas, hasta gaseosas accesibles para el consumo masivo, disponibles en diversos tamaños y tipos de envases. 

La compañía opera en múltiples regiones y zonas, sirviendo a una variedad de puntos de venta que incluyen desde tiendas de conveniencia y minimarkets hasta el canal frío tradicional. Cada tipo de cliente tiene sus particularidades: algunos reciben entregas hasta 4 veces por semana, mientras que otros son visitados por la fuerza de ventas solo una vez semanalmente. Esta diversidad de perfiles representa tanto una oportunidad como un desafío comercial: ¿cómo saber qué productos tienen más chances de ser comprados por cada cliente en un momento dado?

Con el objetivo de aumentar la facturación de forma inteligente y mejorar la eficiencia de su estrategia de ventas, **SodAI Drinks** decide crear una nueva célula interna de innovación: el equipo **Deep Drinkers 🤖**, cuyo propósito es aplicar ciencia de datos para anticiparse a las necesidades del cliente y potenciar el negocio desde una perspectiva basada en información.

El corazón de esta iniciativa es el desarrollo de un sistema predictivo personalizado para cada cliente. Para ello, **Deep Drinkers** convoca a un equipo de Data Scientists y especialistas en *machine learning* con una misión clara: construir un modelo predictivo que, cada semana, pueda estimar la probabilidad de compra de cada producto del portafolio para cada cliente activo.

El modelo deberá tener en cuenta múltiples factores, incluyendo:
- **Tipo de cliente**, ej. "TIENDA DE CONVENIENCIA", "MINIMARKET".
- **Frecuencia de entregas y visitas**, indicadores del nivel de actividad comercial.
- **Ubicación geográfica** (por región y zona).
- **Preferencias históricas de consumo**, inferidas por patrones de compra anteriores.
- **Características del producto**, como marca, categoría, segmento, tipo de envase y tamaño

El objetivo final es que, **cada semana**, se genere una tabla de productos priorizados: para cada cliente, un listado de productos ordenado por su probabilidad estimada de compra. Esta información será enviada al equipo comercial, que podrá usarla en call center, para incrementar las chances de concretar ventas al ofrecer justo lo que el cliente probablemente quiere comprar.

Este proyecto representa un cambio de paradigma en la forma en que **SodAI Drinks** gestiona su fuerza de ventas: de un enfoque reactivo y generalista, a uno proactivo, basado en datos y profundamente personalizado. Así, la empresa no solo espera aumentar su rentabilidad, sino también construir relaciones más sólidas con sus clientes, ofreciéndoles recomendaciones más relevantes y oportunas.

Para lograr lo anterior, el equipo **Deep Drinkers** contará con los siguientes conjuntos de datos, junto a sus respectivos atributos:

- **Datos transaccionales** (`transacciones.parquet`): contiene el historial de compras realizadas por los clientes.
	- `customer_id`: identificador único del cliente que realizó la compra.
	- `product_id`: identificador único del producto comprado.
	- `purchase_date`: fecha en que se realizó la transacción.
	- `order_id`: identificar de la orden de su pedido.
	- `payment`	monto total pagado por la transacción.

- **Datos de clientes** (`clientes.parquet`): incluye las características de cada cliente.
	- `customer_id`: identificador único del cliente.
	- `region_id`: identificador de la región geográfica donde se encuentra el cliente.
	- `customer_type`: tipo de cliente según el canal comercial, por ejemplo, “TIENDA DE CONVENIENCIA”.
	- `Y`: coordenada geográfica de latitud.
	- `X`: coordenada geográfica de longitud.
	- `num_deliver_per_week`: cantidad de entregas semanales que recibe el cliente.
	- `num_visit_per_week`: frecuencia de visitas de la fuerza de ventas por semana.

- **Datos de productos** (`productos.parquet`): describe las características de los productos del portafolio.
	- `product_id`: identificador único del producto.
	- `brand`: marca comercial del producto.
	- `category`: categoría general del producto, como “BEBIDAS CARBONATADAS”.
	- `sub_category`: subcategoría dentro de la categoría principal, por ejemplo, “GASEOSAS”.
	- `segment`: segmento de mercado al que pertenece el producto, como “PREMIUM”.
	- `package`: tipo de envase del producto.
	- `size`: tamaño del producto en litros.

## 📚 Reglas

<center>
<img src="https://media1.tenor.com/m/0Qtv_cQ4ITsAAAAd/necohaus-grey-name.gif" width="450">



El proyecto consta de **dos entregas parciales** y una **entrega final** en donde la primera entrega la idea es poder reflejar lo aprendido durante la primera mitad del curso, que será sobre los contenidos relacionados a *machine learning*, la segunda será sobre los contenidos de la segunda mitad del curso relacionados a *MLOps* y por último la entrega final constará de dos partes, donde la primera será relacionada con experimentación sobre nuevos datasets que serán disponibilizados durante las últimas semanas del curso de manera incremental y una segunda parte que será el informe final escrito que deberá explicar el desarrollo del proyecto completo, como tambien los resultados y análisis de los experimentos realizados sobre los datasets incrementales. La idea es que todo el código esté desarrollado durante las primeras dos entregas y luego en la entrega final sólo se ejecute el código sobre nuevos conjuntos de datos.

La idea de generar el proyecto por etapas es poder aliviar la carga de trabajo en las últimas semanas del semestre donde sabemos que están muy cargado con entregas, pruebas y exámenes de otros ramos, y así garantizamos que habiendo la desarrollado las dos primeras entregas parciales, tendrán el grueso del proyecto listo para luego experimentar y documentar.

---
### **Fechas de entrega**
- **Entrega parcial 1**: 14 de Mayo
- **Entrega parcial 2**: Por definir
- **Entrega final**: Por definir

---

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. No se aceptarán trabajos individuales o grupos con más integrantes.
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

### **Entregables y etapas**

#### **1. Entrega Parcial 1**  
- Dispondrán de los archivos de datos **productos.parquet**, **clientes.parquet** y **transacciones.parquet** para el modelamiento inicial.  
- Utilizarán estos archivos para desarrollar lo solicitado para la entrega 1. 
- En esta etapa, se espera que apliquen todos los conocimientos aprendidos durante la primera parte del curso relacionados con *machine learning*.
- **Informe**: No se exige un avance del informe en esta etapa, sólo un notebook con su desarrollo actual, pero se **recomienda comenzar** a redactar el informe final en paralelo para disminuir la carga académica en las etapas posteriores.  

#### **2. Entrega Parcial 2**  
- En esta entrega, deberán aplicar los conocimientos aprendidos durante la segunda mitad del curso sobre *MLOps*  
- Se espera que implementen estos conocimientos para desplegar su modelo elegido en la primera entrega y crear *pipelines* automatizados que simulen un entorno productivo.
- **Informe**: similar a la primera etapa, no se exige un avance del informe, pero se **recomienda avanzar con su redacción** para evitar una acumulación de trabajo en la etapa final.  

#### **3. Entrega Final**  
- En la entrega final, deberán realizar dos etapas:
	- La primera etapa es sobre experimentación utilizando datasets incrementales que se irán disponibilizando de manera parcial, para que vayan generando predicciones con su modelo ya desplegado. El objetivo de esta etapa es poder testear su solución *end-to-end* y que vayan analizando los resultados obtenidos a medida que se van agregando más datos.
	- La segunda etapa consiste en redactar un informe final que deberá explicar el desarrollo completo de tu proyecto y un análisis profundo de sus resultados de experimentación. Este informe debera incluir a lo menos las siguientes secciones:
		- Análisis exploratorio de datos  
		- Metodología aplicada  
		- Selección y entrenamiento de modelos  
		- Evaluación de resultados  
		- Optimización de modelos
		- Interpretabilidad
		- Re-entrenamiento
		- Tracking con MLFlow
		- Creación de la aplicación web con Gradio y FastAPI

Es **altamente recomendable** ir redactando el informe en paralelo al desarrollo de los modelos para garantizar que toda la información relevante quede documentada adecuadamente.  

### Nota Final

La calificación final de su proyecto se calculará utilizando la siguiente ponderación: 

$$Nota Final = 0.30 * EntregaParcial1 + 0.40 * EntregaParcial2 + 0.30 * EntregaFinal$$

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)

# 📬 Entrega Parcial 1 (30% del Proyecto)

### 📪 Fecha de Entrega: 14 de Mayo

## 📌 Abstract [0.25 puntos]

<center>
<img src="https://i.redd.it/h5ptnsyabqvd1.gif" width="400" height="300">

En esta sección, deben redactar un Abstract claro y conciso para su proyecto. El Abstract debe responder a las siguientes preguntas clave:

- **Descripción del problema**: ¿Cuál es el objetivo del proyecto? ¿Qué se intenta predecir o analizar?
- **Datos de entrada**: ¿Qué datos tienen disponibles? ¿Cuáles son sus principales características?
- **Métrica de evaluación**: ¿Cómo medirán el desempeño de sus modelos? Expliquen por qué eligieron esta métrica basándose en el análisis exploratorio de los datos.
- **Modelos y transformaciones**: ¿Qué modelos utilizarán y por qué? ¿Qué transformaciones o preprocesamientos aplicaron a los datos?
- **Resultados generales**: ¿El modelo final cumplió con los objetivos del proyecto? ¿Cuáles fueron las conclusiones más importantes?

**Importante**: Escriban esto despues de haber resuelto el resto de la tarea.

> [Escriba aquí su Abstract]

## 📌 Pre-procesamiento [0.5 puntos]

<center>
<img src="https://media0.giphy.com/media/10zsjaH4g0GgmY/giphy.gif?cid=6c09b9523xtlunksc9amikw09zk1bmiqwjqnt70ae82rk877&ep=v1_gifs_search&rid=giphy.gif&ct=g" width="400" height="300">

Tal como en muchos otros problemas de negocio, los datos probablemente deben ser pre procesados antes de aplicar cualquier técnica de analítica. Bajo esa premisa, en esta sección deben desarrollar código que les permita **preparar los datos** de tal forma que les permita resolver el problema planteado. Para esto, pueden aplicar procesamientos como:

- Transformaciones de tipo de dato (str, int, etc)
- Cruce de información
- Eliminación de duplicados
- Filtros de fila y/o columnas

*Hint: ¿Qué forma debería tener la data para resolver un problema de aprendizaje supervisado?*

Todo proceso llevado a cabo debe estar bien documentado y justificado en el informe, explicando el por qué se decidió realizar en funcion de los datos presentados y los objetivos planteados del proyecto.

### Clientes

In [5]:
import pandas as pd

clients = pd.read_parquet("clientes.parquet")
clients.head()

customer_id  region_id  zone_id customer_type          Y           X  \
10705       256017         80     5148     ABARROTES -46.474800 -108.045140   
10706       255780         80     5148     ABARROTES -46.520282 -107.961052   
10707       254655         80     5148     ABARROTES -46.537640 -107.909280   
10708       254445         80     5148     ABARROTES -46.543526 -107.917954   
10709       254403         80     5148     ABARROTES -46.540856 -107.905792   

       num_deliver_per_week  num_visit_per_week  
10705                     1                   1  
10706                     1                   1  
10707                     1                   1  
10708                     2                   1  
10709                     3                   1

In [24]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1569 entries, 10705 to 12273
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           1569 non-null   int64  
 1   region_id             1569 non-null   int64  
 2   zone_id               1569 non-null   int64  
 3   customer_type         1569 non-null   object 
 4   Y                     1569 non-null   float64
 5   X                     1568 non-null   float64
 6   num_deliver_per_week  1569 non-null   int64  
 7   num_visit_per_week    1569 non-null   int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 110.3+ KB


Los tipos de las variables estan acorde a lo que representan y notamos un único valor nulo para la coordenada X

In [ ]:
# inspeccionar duplicados:
clients[clients.duplicated()]

Empty DataFrame
Columns: [customer_id, region_id, zone_id, customer_type, Y, X, num_deliver_per_week, num_visit_per_week]
Index: []

No tenemos valores duplicados

### Productos

In [6]:
products = pd.read_parquet("productos.parquet")
products.head()

product_id     brand              category sub_category  segment  package  \
0       34092  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
1       57290  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
2       56714  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
3      296616  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
4       60854  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   

   size  
0  0.66  
1  0.66  
2  0.66  
3  0.66  
4  0.25

In [27]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 971 entries, 0 to 970
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    971 non-null    int64  
 1   brand         971 non-null    object 
 2   category      971 non-null    object 
 3   sub_category  971 non-null    object 
 4   segment       971 non-null    object 
 5   package       971 non-null    object 
 6   size          971 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 53.2+ KB


No tenemos valores nulos y los tipos de las columnas corresponden.

In [28]:
products['size'].describe()

count    971.000000
mean       1.420525
std        3.897778
min        0.125000
25%        0.310000
50%        0.500000
75%        1.000000
max       20.000000
Name: size, dtype: float64

In [30]:
products[products.duplicated()]

Empty DataFrame
Columns: [product_id, brand, category, sub_category, segment, package, size]
Index: []

No tenemos valores duplicados

### Transacciones

In [7]:
transactions = pd.read_parquet("transacciones.parquet")
transactions.head()

customer_id  product_id  order_id purchase_date     items
124        61353       61364    411145    2024-04-27 -0.333333
127        61899        1370    411156    2024-04-27  3.666667
344       154077       30500    417911    2024-04-29  2.333333
349       164337       56714    418100    2024-04-29 -2.333333
354       172056       61672    407162    2024-04-26  1.000000

In [31]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 254936 entries, 124 to 7712548
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   customer_id    254936 non-null  int64         
 1   product_id     254936 non-null  int64         
 2   order_id       254936 non-null  int64         
 3   purchase_date  254936 non-null  datetime64[ns]
 4   items          254936 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 11.7 MB


In [32]:
products.loc[products['product_id'] == 61364]

product_id     brand              category sub_category segment  package  \
462       61364  Brand 35  BEBIDAS CARBONATADAS     GASEOSAS  MEDIUM  BOTELLA   

     size  
462   1.0

In [33]:
transactions['items'].describe()

count    254936.000000
mean          4.378552
std          12.098818
min        -399.666667
25%           1.000000
50%           2.333333
75%           3.666667
max        1000.333333
Name: items, dtype: float64

Notamos que no hay valores nulos. La columna items parece no tener mucho sentido que sea de tipo float, ya que no se puede comprar una cantidad decimal de productos como gaseosas. Además notamos que hay cantidades negativas en items.

In [35]:
products['category'].unique() # asegurar que solo sean bebidas

array(['BEBIDAS CARBONATADAS', 'BEBIDAS NO CARBONATADAS'], dtype=object)

In [39]:
# valores duplicados:
transactions[transactions.duplicated()]

customer_id  product_id  order_id purchase_date  items
7217          212550       61364    419150    2024-04-29      0
14464         217782       57804    419391    2024-04-29      7
26103         217782       50716    419391    2024-04-29     33
32338         197286       11262    418607    2024-04-29      1
42356          61353       11262    411145    2024-04-27      0
...              ...         ...       ...           ...    ...
7691409       154551       11262    504913    2024-06-03      1
7696763       212550       61364    505599    2024-06-03      1
7697616       176643       56714    505052    2024-06-03      1
7704734       188196       61364    489536    2024-05-27      1
7709868       207339       59816    489756    2024-05-27      1

[1215 rows x 5 columns]

In [40]:
# eliminar duplicados:
transactions = transactions.drop_duplicates()

In [41]:
transactions[transactions['items'] < 0]

customer_id  product_id  order_id purchase_date  items
349           164337       56714    418100    2024-04-29     -2
359           176643       56714    418196    2024-04-29     -2
362           181593       56714    418241    2024-04-29     -2
363           183213       61348    418256    2024-04-29     -1
367           184698        1370    418277    2024-04-29     -1
...              ...         ...       ...           ...    ...
7703661       156489       30500    489422    2024-05-27     -1
7707199       205821       57414    489737    2024-05-27     -1
7707738       205821       61542    489737    2024-05-27     -1
7709645       205821       56714    489737    2024-05-27     -1
7710449       205821       61364    489737    2024-05-27     -1

[6087 rows x 5 columns]

Aca decidir si se borran o no 6087 filas y si tiene sentido cambiar a int

In [37]:
# convertir a int64 los items:
transactions['items'] = transactions['items'].astype('int64')
transactions.head()

customer_id  product_id  order_id purchase_date  items
124        61353       61364    411145    2024-04-27      0
127        61899        1370    411156    2024-04-27      3
344       154077       30500    417911    2024-04-29      2
349       164337       56714    418100    2024-04-29     -2
354       172056       61672    407162    2024-04-26      1

TO-DO:

- Cruzar información
- Checkear si al cruzar son los mismos clientes y productos que en transactions
- Filtros de columnas
- Decidir que hacer con columna items

## 📌 EDA [0.5 puntos]

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExbHZ6aGdkd21tYTI3cW8zYWhyYW5wdGlyb2s3MmRzeTV0dzQ1NWlueiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3k1hJubTtOAKPKx4k3/giphy.gif" width="400" height="200">

En esta sección, se debe realizar un análisis exploratorio de los datos para comprender su estructura, detectar posibles problemas y obtener información relevante para el entrenamiento de los modelos. La idea es que puedan detectar **patrones en los datos** que les permitan resolver el problema con mayor facilidad.

Se deben responder preguntas a partir de lo que puedan visualizar/obtener, por ejemplo:

- Clientes y productos

    - ¿Cuántos clientes únicos hay en el dataset?

    - ¿Cuántos productos únicos se encuentran en los datos?

- Periodo y frecuencia

    - ¿De qué periodo es la información disponible?

    - ¿Cuál es la frecuencia de los registros (diaria, semanal, mensual, etc.)?

- Calidad de los datos

    - ¿Existen valores nulos en el dataset? ¿Cuántos? ¿Cómo se pueden tratar?

    - ¿Hay datos raros, como cantidades negativas o inconsistencias? Genere tests de validación para identificar estos problemas.

- Patrones de compra

    - ¿Cuántos productos compra en promedio cada cliente semana a semana?

    - ¿Cuántas transacciones ha realizado cada cliente?

    - ¿Cuál es el periodo de recompra promedio de cada SKU?

In [42]:
# 1. Valores únicos
print(f"{len(transactions)} transacciones en total")
print(f"{transactions['order_id'].nunique()} órdenes")

print("\nClientes:")
print(f"{clients['customer_id'].nunique()} clientes únicos")
print(f"{transactions['customer_id'].nunique()} clientes con transacciones")

print("\nProductos:")
print(f"{products['product_id'].nunique()} productos únicos")
print(f"{transactions['product_id'].nunique()} productos en transacciones")


253721 transacciones en total
64600 órdenes

Clientes:
1569 clientes únicos
1490 clientes con transacciones

Productos:
971 productos únicos
114 productos en transacciones


In [44]:
# 2. Periodo y frecuencia
print(f"Periodo de tiempo: {transactions['purchase_date'].min()} - {transactions['purchase_date'].max()}")

# TO-DO: calcular la frecuencia de los registros

Periodo de tiempo: 2024-01-01 00:00:00 - 2024-12-31 00:00:00


In [ ]:
# 3. Calidad de los datos
# TO-DO: implementar

In [ ]:
# 4. Patrones de compra
client_transactions = transactions.groupby('customer_id').agg('count').reset_index()
client_transactions = client_transactions[['customer_id', 'items']]
client_transactions.columns = ['customer_id', 'transactions_count']
client_transactions.sort_values('transactions_count', ascending=False, inplace=True)
client_transactions.head(10)

customer_id  transactions_count
416       183105                1596
570       197286                1303
456       186174                1281
138       153303                1206
492       188331                1189
2          25749                1181
483       187806                1108
568       196845                1094
545       194229                1073
98         72930                1068

In [97]:
client_transactions['transactions_count'].describe()

count    1490.000000
mean      170.282550
std       182.331907
min         1.000000
25%        55.000000
50%       114.000000
75%       217.000000
max      1596.000000
Name: transactions_count, dtype: float64

In [ ]:
transactions['week'] = transactions['purchase_date'].dt.isocalendar().week
transactions.loc[transactions['purchase_date'] > '2024-12-29', 'week'] = 53 # caso borde

In [91]:
transactions.columns

Index(['customer_id', 'product_id', 'order_id', 'purchase_date', 'items',
       'week'],
      dtype='object')

In [93]:
products_per_week = transactions.groupby(['customer_id', 'week']).agg({'items': 'sum'}).reset_index()
products_per_week

customer_id  week  items
0            25734     1      9
1            25734     2     18
2            25734     4     28
3            25734     5     15
4            25734     6     20
...            ...   ...    ...
46424      2061063    44      6
46425      2061063    46      6
46426      2061063    47     27
46427      2061063    51      8
46428      2061063    52      9

[46429 rows x 3 columns]

In [95]:
client_products_per_week = products_per_week.groupby('customer_id').agg({'items': 'mean'}).reset_index()
client_products_per_week.columns = ['customer_id', 'avg_items_per_week']
client_products_per_week.sort_values('avg_items_per_week', ascending=False, inplace=True)
client_products_per_week.head(10)

customer_id  avg_items_per_week
401        181101          766.428571
492        188331          522.764706
483        187806          384.274510
1439       250434          323.363636
294        164064          317.480000
778        206748          317.037736
963        211407          314.600000
369        176403          285.235294
214        154890          280.098039
344        172161          271.078431

In [96]:
client_products_per_week['avg_items_per_week'].describe()

count    1490.000000
mean       17.407667
std        41.395028
min         0.666667
25%         3.763736
50%         6.710084
75%        14.410675
max       766.428571
Name: avg_items_per_week, dtype: float64

In [ ]:
# TO-DO: Implementar periodo de recompra promedio para cada producto (SKU)

## 📌 Holdout [0.25 puntos]

Para evaluar correctamente el modelo y garantizar su capacidad de generalización, se deben dividir los datos en tres conjuntos: 
- `Entrenamiento` : Para ajustar los parámetros.
- `Validación`: Para optimizar hiperparámetros y seleccionar el mejor modelo.
- `Prueba` : Para evaluar el rendimiento final en datos no vistos.

👀 **Hint**: *Recuerde que los datos tienen una temporalidad que debe considerarse al momento de separarlos, para evitar fugas de información. Es importante justificar la estrategia de partición elegida y visualizar la distribución temporal de los conjuntos generados*

In [11]:
# Desarrolle aquí su código

## 📌 Feature Engineering [0.5 puntos]

<center>
<img src="https://i.imgur.com/CmXZSSC.gif" width="300" height="200">

En esta sección, se deben construir pipelines para automatizar el preprocesamiento de los datos, lo cual garantizará que el flujo de trabajo sea reproducible y eficiente para esta entrega y las futuras. El objetivo es aplicar una serie de transformaciones en un orden definido para asegurar que los datos estén listos para los modelos a entrenar. El pipeline final debe incluir las técnicas de pre-procesamiento que se deben aplicar a los distintos datos (según lo que consideren necesario para el problema). Por ejemplo:

- **Imputación de valores nulos**: Manejo de datos faltantes mediante estrategias adecuadas (media, mediana, moda, interpolación, etc.). 

- **Transformaciones personalizadas**: Uso de ColumnTransformer para aplicar diferentes transformaciones a columnas específicas.

- **Codificación de variables categóricas**: Convertir datos categóricos a un formato numérico adecuado (One-Hot Encoding, Label Encoding, etc.).

- **Discretización de variables**: Conversión de variables numéricas continuas en categorías si son relevantes para el desempeño del modelo a entrenar.

- **Estandarización o normalización** : Ajustar la escala de los datos para mejorar el rendimiento de los algoritmos sensibles a la magnitud de las variables.

- **Eliminación o transformación de valores atípicos**: Identificar y tratar con datos outliers para mejorar la robustez del modelo.

- **Nuevas características** : Creación de variables adicionales que puedan aportar información relevante al modelo.

Cada una de estas transformaciones debe ser justificada en función de su relevancia para el problema y los datos, y es importante evaluar su impacto en el rendimiento del modelo. Además, el pipeline debe ser flexible y modular para poder probar diferentes configuraciones de preprocesamiento.

In [13]:
# Desarrolle aquí su código 

## 📌 Baseline [0.25 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExN3lzeGFqZmU3NzJrZHllNjRmaHVzczJpZ29rdHdlMzVpZnQwNXo1diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/qAtZM2gvjWhPjmclZE/giphy.gif" width="300" height="200">

En esta sección se debe construir el modelo más sencillo posible que pueda resolver el problema planteado, conocido como **Modelo baseline**. Su propósito es servir como referencia para comparar el rendimiento de los modelos más avanzados desarrollados en etapas posteriores.  

Pasos requeridos:  
- Implemente, entrene y evalúe un modelo básico utilizando un pipeline.  
- Asegúrese de incluir en el pipeline las transformaciones del preprocesamiento realizadas previamente junto con un clasificador básico.  
- Evalúe el modelo y presente el informe de métricas utilizando **`classification_report`**.  

Documente claramente cómo se creó el modelo, las decisiones tomadas y los resultados obtenidos. Este modelo será la base comparativa en las secciones posteriores.

In [15]:
# Desarrolle aquí su código 

## 📌 Elección de modelo [0.75 puntos]

En esta sección deben escoger un modelo que se adapte a las necesidades del negocio. Para esto, pruebe al menos 3 modelos y desarrolle los siguientes aspectos para cada uno:

- **Estructura y diferencias entre los modelos**: Explicar brevemente cada uno y sus hipérparámetros de mayor importancia.
- **Clasificadores recomendados**:
  - `LogisticRegression`
  - `KNeighborsClassifier`
  - `DecisionTreeClassifier`
  - `SVC`
  - `RandomForestClassifier`
  - `LightGBMClassifier` (del paquete `lightgbm`)
  - `XGBClassifier` (del paquete `xgboost`)
  - Otro (según lo que se estime adecuado)
  
- **Evaluación de resultados**: Se utilizará el **`classification_report`** para evaluar el rendimiento de cada modelo, destacando métricas clave como precisión, recall y F1-score. **Importante: No optimicen hiperparámetros, la idea es hacer una selección rápida del modelo.**

**Nota:** Pueden ocupar mas de 1 **instancia** de modelo para resolver el problema (e.g: (modelo_1, grupo_1), (modelo_2, grupo_2), ...).
  
A continuación, se deben responder las siguientes preguntas para evaluar el rendimiento de los modelos entrenados:

1. ¿Hay algún clasificador que supere al modelo baseline?  
2. ¿Cuál es el mejor clasificador entrenado y por qué?  
3. ¿Qué factores hacen que el mejor clasificador sea superior a los otros?  
4. En términos de `tiempo de entrenamiento`, ¿Qué modelo considera más adecuado para experimentar con grillas de optimización?

In [17]:
# Desarrolle aquí su código 

## 📌 Optimización de Hiperparámetros [1.0 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExcXJkNzdhYjlneHplaGpsbnVkdzh5dnY3Y2VyaTIzamszdGR1czJ2diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/2rqEdFfkMzXmo/giphy.gif" width="300" height="200">

A partir de su análisis anterior, se debe proceder a optimizar el rendimiento del modelo seleccionado mediante la optimización de sus hiperparámetros. Para ello, se espera que implementen `Optuna` para optimizar no solo los hiperparámetros del modelo, sino también los de los preprocesadores utilizados (por ejemplo, OneHot Encoding, Scalers, etc.).

Al desarrollar este proceso, deberán responder las siguientes preguntas clave como mínimo:

- ¿Qué métrica decidieron optimizar y por qué?

- ¿Qué hiperparámetro tuvo un mayor impacto en el rendimiento de su modelo?

- ¿Cuánto mejoró el rendimiento del modelo después de la optimización de hiperparámetros?

In [20]:
# Desarrolle aquí su código

## 📌 Interpretabilidad [1.0 puntos]

En esta sección, deben explicar el funcionamiento de su modelo utilizando las técnicas de interpretabilidad vistas en clase, como `SHAP`. Se espera que sean capaces de descomponer las predicciones y evaluar la importancia de los atributos y las interacciones entre ellos, con el fin de obtener una comprensión más profunda de cómo el modelo toma decisiones. 

Al desarrollar esta parte, deberán responder las siguientes preguntas clave como mínimo:

- ¿Podría explicar el funcionamiento de su modelo para una predicción en particular? Si es así, proporcione al menos tres ejemplos específicos, describiendo cómo el modelo llegó a sus decisiones y qué factores fueron más relevantes en cada caso.

- ¿Qué atributo tiene una mayor importancia en la salida de su modelo? Analice si esto tiene sentido con el problema planteado y justifique la relevancia de dicho atributo en el contexto de las predicciones que se realizan.

- ¿Existe alguna interacción entre atributos que sea relevante para el modelo? Investigue si la combinación de ciertos atributos tiene un impacto significativo en las predicciones y explíquela en **detalle**.

- ¿Podría existir sesgo hacia algún atributo en particular? Reflexione sobre la posibilidad de que el modelo esté favoreciendo ciertos atributos. Si es así, ¿cuál podría ser la causa y qué impacto podría tener esto en la predicción?

In [21]:
# Desarrolle aquí su código

## 📌 Resultados y Conclusiones [1.0 puntos]

Para finalizar, se deben explicar los desarrollos y resultados obtenidos a lo largo de todo el proceso, desde la selección de las variables hasta la optimización de hiperparámetros e interpretación. Se espera una reflexión crítica sobre el desempeño de los modelos entrenados y una comparación entre los diferentes enfoques. Además, deberán abordar los siguientes puntos clave:

- **Análisis de métricas**: Comenten sobre las métricas obtenidas en cada etapa del modelo, destacando las más relevantes como precisión, recall, F1-score, etc. ¿Cuáles fueron los modelos más efectivos? ¿Qué diferencias notables encontró entre ellos?

- **Impacto de las decisiones tomadas**: Reflexionen sobre cómo las decisiones relacionadas con el preprocesamiento, selección de atributos y optimización de hiperparámetros influyeron en los resultados finales. ¿Hubo alguna decisión que haya tenido un impacto notable en el rendimiento?

- **Lecciones aprendidas**: Concluyan sobre las lecciones más importantes que aprendieron durante el proceso y cómo estas pueden influir en futuras iteraciones del modelo. ¿Qué se podría mejorar si se repitiera el proceso? Si tuvieran más recursos y tiempo, ¿qué otras técnicas/herramientas habrían utilizado?

> [Escriba aquí sus resultados]

Mucho éxito!

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHpvOTY5Z3hpdHI3aDBpdGRueXRqamZncXp2emFrbjJ5M2s5eTR1dSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1PMVNNKVIL8Ig/giphy.gif" width="300" height="200">
